platform: mtl: ipc fw_ready fix

aborisovich opened the pull request:
IPC boot complete - partial fix, partial refactoring. Sending IPC should use high level APIs and functions.
Now ipc header is initialized on heap and filled with zeroes for safety.

Signed-off-by: Andrey Borisovich <andrey.borisovich@intel.com>

softwarecki commented on src/platform/intel/ace/platform.c:
```diff
@@ -52,6 +54,16 @@ static const struct sof_ipc_fw_ready ready
+	mailbox_sw_reg_write(SRAM_REG_FW_STATUS, p & 0x3fffffff);
+	struct ipc_cmd_hdr header;
+	header.pri = 0x80000000 | (p & 0x3fffffff);
+	header.ext = MAILBOX_EXCEPTION_OFFSET + 2 * 0x20000;
```
No magic values please

aborisovich replied:
Same is done for apollolake, cannonlake, icelake... Just following the convention :-)

softwarecki commented on src/drivers/intel/ace/ipc.c:
```diff
@@ -169,5 +170,7 @@ int platform_ipc_init(struct ipc *ipc)
 	/* attach handlers */
 	intel_adsp_ipc_set_message_handler(INTEL_ADSP_IPC_HOST_DEV, message_handler, ipc);
 
+	irq_enable(PLATFORM_IPC_INTERRUPT);
```
It will fail during compilation. PLATFORM_IPC_INTERRUPT isn't defined for mtl platform. Get interrupt number from device tree.

aborisovich replied:
> It will fail during compilation. PLATFORM_IPC_INTERRUPT isn't defined for mtl platform. Get interrupt number from device tree.

They are all in the soc\xtensa\intel_adsp\ace_v1x\include\ace_v1x-regs.h header (not yet in the devicetree) but retrieved them using proper macros 👍

softwarecki replied:
Have you ever tried compiling this code? :joy:

kv2019i commented on src/platform/intel/ace/platform.c:
```diff
@@ -52,6 +54,18 @@ static const struct sof_ipc_fw_ready ready
 	.flags = DEBUG_SET_FW_READY_FLAGS,
 };
 
+void platform_panic(uint32_t p)
```
Very welcome patch @aborisovich , this is meeting https://github.com/thesofproject/sof/issues/6196 . But, but, is this called, part of the problem is how to hook this (fairly app specific IPC) to Zephyr panic handling. Do you have a solution to this somewhere?

aborisovich replied:
No I don't thought about it. This is purely SOF solution...

aborisovich replied:
Dropped platform_panic from this PR.

aborisovich commented on src/platform/intel/ace/platform.c:
```diff
@@ -20,6 +20,8 @@
 #include <ipc/info.h>
 #include <kernel/abi.h>
+#include <sof/ipc/msg.h>
+#include <ipc/header.h>
```
?

aborisovich commented on src/drivers/intel/ace/ipc.c:
```diff
@@ -169,5 +170,7 @@ int platform_ipc_init(struct ipc *ipc)
 	/* attach handlers */
 	intel_adsp_ipc_set_message_handler(INTEL_ADSP_IPC_HOST_DEV, message_handler, ipc);
 
+	irq_enable(MTL_IRQ_TO_ZEPHYR(MTL_INTL_HIPC));
```
No. IPC driver implementation is infact in SOF, Zephyr only provides API for driver to use
(I mean zephyr\soc\xtensa\intel_adsp\common\include\intel_adsp_ipc.h).
So if we initialize the driver (and API) in SOF we also enable interrupt service request in SOF where initialization function is.

aborisovich replied:
removed

kv2019i commented on src/platform/intel/ace/platform.c:
```diff
@@ -59,15 +59,11 @@ int platform_boot_complete(uint32_t boot_message)
 	/* send fimrware ready message. */
-	return ipc_platform_send_msg(&msg);
+	ipc_msg_send(ipc_msg, NULL, true);
+	return 0;
```
Hmm, not following. You pass "ipc_msg" to ipc_msg_send() and msg is not used? And 'ready' is not used anywhere here. How can this work?

aborisovich replied:
Please have a look at function definition in src\ipc\ipc-common.c
![image](https://user-images.githubusercontent.com/12482515/194788645-881148ad-1751-4084-97a7-a84a59896e22.png)
@kv2019i please alaborate on what 'ready' is?
If you ask why `platform_boot_complete` function argument `uint32_t boot_message` is unused, it is because function signature must match generic for all platforms src\include\sof\platform.h declaration.
@softwarecki I'm glad your colleagues work puts you in a good mood.

aborisovich replied:
Ah I see now, was renaming variables really fast and forgot this one. Just a typo.

kv2019i reviewed (changes requested):
I might be missing something, but please see comment inline.

aborisovich commented:
Update:

- dropped commit with platform_panic as adding callback from Zephyr mtrace API does not look trivial to do but it is preferable way
- fixed code in platform.c with `platform_boot_complete`

lgirdwood reviewed (commented):
Are we using the pre-allocated IPC memory here or a new heap allocation ? Just a reminder if its a new allocation I cant see a free().

softwarecki commented:
@lgirdwood ```ipc_msg_w_ext_init``` dynamically allocates memory for the message structure and separately for the payload. I think @aborisovich should not set a pointer to the static const struct to the tx_data.

aborisovich commented:
> @lgirdwood ipc_msg_w_ext_init dynamically allocates memory for the message structure and separately for the payload. I think @aborisovich should not set a pointer to the static const struct to the tx_data.

Why not? I've allocated nothing for the message data (tx_data) because it's allocated in static memory.
I don't think it is possible to allocate heap memory for tx_data that is expected to be found in `__section(".fw_ready")`?
